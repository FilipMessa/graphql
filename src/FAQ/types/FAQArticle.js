// @flow

// this is fetched from "/categories/:id" endpoint
export type RawFAQArticleItem = {
  url: string,
  title: string,
  perex: string,
  upvotes: number,
  downvotes: number,
};

export type SanitizedFAQArticleItem = RawFAQArticleItem & {
  id: string,
  language: string,
};

// this is fetched from "/articles/:id" endpoint
export type FAQArticleDetail = {
  id: number,
  content: string,
  title: string,
  perex: string,
  upvotes: number,
  downvotes: number,
};

export type FAQArticleResponse = SanitizedFAQArticleItem & FAQArticleDetail;
