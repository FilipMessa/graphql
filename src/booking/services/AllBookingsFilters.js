// @flow

import idx from 'idx';
import { DateTime } from 'luxon';

import type { BookingsItem } from '../Booking';
import type { Leg } from '../../flight/Flight';
import { CoveredBy } from '../../flight/types/enums/CoveredBy';

export function groupBookings(bookings: $ReadOnlyArray<BookingsItem>) {
  return bookings.reduce(
    (acc, curVal) => {
      if (isPastBooking(curVal) === true) {
        acc.past.push(curVal);
      } else {
        acc.future.push(curVal);
      }
      return acc;
    },
    {
      future: [],
      past: [],
    },
  );
}

export function filterOnlyBookings(
  timeFrame: 'future' | 'past',
  bookings: $ReadOnlyArray<BookingsItem>,
) {
  const bookingGroups = groupBookings(bookings);

  if (timeFrame === 'future') {
    return bookingGroups.future;
  }
  return bookingGroups.past;
}

/**
 * Past booking is booking where every arrival date of every leg is in past
 * (considering local time).
 */
export function isPastBooking(singleBooking: BookingsItem) {
  const when = singleBooking.arrival.when;
  if (!when) {
    return true;
  }

  return (
    DateTime.fromJSDate(when.utc, {
      zone: 'utc',
    })
      .diffNow('milliseconds')
      .toObject().milliseconds < 0
  );
}

export function findUpcomingLeg(
  legs: Leg[],
  guarantee: ?$Values<typeof CoveredBy>,
) {
  return [...legs]
    .filter(leg => {
      const date = idx(leg, _ => _.arrival.when.utc);
      const isFuture = date
        ? DateTime.fromJSDate(date, { zone: 'utc' }).diffNow('milliseconds')
            .milliseconds > 0
        : false;

      return guarantee ? leg.guarantee === guarantee && isFuture : isFuture;
    })
    .sort((legA, legB) => {
      const dateA = idx(legA, _ => _.arrival.when.utc);
      const dateB = idx(legB, _ => _.arrival.when.utc);

      return dateA && dateB ? dateA - dateB : 0;
    })
    .shift();
}
