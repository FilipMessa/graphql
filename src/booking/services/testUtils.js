// @flow

export const createLeg = (milliseconds: number) => ({
  arrival: {
    when: {
      utc: new Date(milliseconds),
    },
  },
});
