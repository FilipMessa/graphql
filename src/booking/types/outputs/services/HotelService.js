// @flow

import { GraphQLObjectType, GraphQLList } from 'graphql';

import HotelServiceRelevantLocation from './HotelServiceRelevantLocation';

type AncestorType = {|
  +relevantLocationCodes: $ReadOnlyArray<string>,
|};

export default new GraphQLObjectType({
  name: 'HotelService',
  fields: {
    relevantLocations: {
      type: GraphQLList(HotelServiceRelevantLocation),
      resolve: ({ relevantLocationCodes }: AncestorType) =>
        relevantLocationCodes,
    },
  },
});
