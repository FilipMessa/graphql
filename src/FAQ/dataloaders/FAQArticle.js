// @flow

import Dataloader from 'dataloader/index';
import stringify from 'json-stable-stringify';

import { get } from '../../common/services/HttpRequest';
import Config from '../../../config/application';

type NotFound = {|
  detail: string,
|};
type ValidResponse = {
  id: number,
  content: string,
  title: string,
  perex: string,
  upvotes: number,
  downvotes: number,
};
type APIResponse = NotFound | ValidResponse;

export type FAQArticleDetail = ValidResponse & {
  language: string,
};

export type Args = {|
  originalId: string,
|};

const FAQArticle = async (
  originalId: string,
  language: string,
): Promise<FAQArticleDetail> => {
  const url = Config.restApiEndpoint.FAQArticle(originalId);
  const article: APIResponse = await get(url, null, {
    'Accept-Language': language,
  });

  if (!article || article.detail) {
    throw new Error(`Article with id ${originalId} was not found.`);
  }

  return {
    ...article,
    language,
  };
};

const batchLoad = async (
  inputs: $ReadOnlyArray<Args>,
  language: string,
): Promise<Array<FAQArticleDetail>> => {
  const promises = inputs.map(({ originalId }: Args) =>
    FAQArticle(originalId, language),
  );

  return Promise.all(promises);
};

export default function createFAQLoader(language: string) {
  return new Dataloader(queries => batchLoad(queries, language), {
    cacheKeyFn: key => stringify(key),
  });
}
