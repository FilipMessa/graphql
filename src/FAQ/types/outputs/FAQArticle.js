// @flow

import { GraphQLInt, GraphQLString, GraphQLObjectType } from 'graphql';

import { globalIdField } from '../../../common/services/OpaqueIdentifier';
import type { GraphqlContextType } from '../../../common/services/GraphqlContext';
import type { FAQArticleDetail } from '../../dataloaders/getFAQArticle';
import type { FAQArticleItem } from '../../dataloaders/getFAQCategory';
import type { ArticleFromSearch } from '../../dataloaders/searchFAQ';

type FAQArticleResponse = FAQArticleItem | FAQArticleDetail | ArticleFromSearch;

export type FAQArticleType = {|
  id: number,
  originalId: number,
  title: string,
  perex: string,
  content: string,
  upvotes: number,
  downvotes: number,
|};

export default new GraphQLObjectType({
  name: 'FAQArticle',
  fields: {
    id: globalIdField('FAQArticle', (response: FAQArticleResponse) =>
      String(response.id),
    ),
    originalId: {
      type: GraphQLInt,
      description: 'Original numeric id of the article',
      resolve: ({ id }: FAQArticleResponse): number => id,
    },
    title: {
      type: GraphQLString,
      description: 'Title of the article',
      resolve: ({ title }: FAQArticleResponse): string => title,
    },
    perex: {
      type: GraphQLString,
      description: 'Perex of the article',
      resolve: ({ perex }: FAQArticleResponse): string => perex,
    },
    content: {
      type: GraphQLString,
      description: 'Content of the article',
      resolve: async (
        response: FAQArticleResponse,
        args: Object,
        { dataLoader }: GraphqlContextType,
      ): Promise<string> => {
        // "content" is present only for FAQArticleDetail
        const article = await dataLoader.FAQArticle.load({
          originalId: String(response.id),
        });
        return article.content;
      },
    },
    upvotes: {
      type: GraphQLInt,
      description: 'Upvotes about the utility of the article',
      resolve: ({ upvotes }: FAQArticleResponse): number => upvotes,
    },
    downvotes: {
      type: GraphQLInt,
      description: 'Downvotes about the utility of the article',
      resolve: ({ downvotes }: FAQArticleResponse): number => downvotes,
    },
  },
});
