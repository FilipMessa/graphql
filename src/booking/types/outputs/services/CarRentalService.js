// @flow

import { GraphQLObjectType, GraphQLList } from 'graphql';

import CarRentalServiceRelevantCities, {
  type CarRentalServiceRelevantCitiesType,
} from './CarRentalServiceRelevantCities';
import { type GraphqlContextType } from '../../../../common/services/GraphqlContext';

export type CarRentalServiceType = Map<
  string,
  {|
    +pickup: Date,
    +dropoff: Date,
  |},
>;

export default new GraphQLObjectType({
  name: 'CarRentalService',
  fields: {
    relevantCities: {
      type: GraphQLList(CarRentalServiceRelevantCities),
      resolve: async (
        ancestor: CarRentalServiceType,
        args: {},
        context: GraphqlContextType,
      ): Promise<CarRentalServiceRelevantCitiesType[]> => {
        const locations = await Promise.all(
          Array.from(ancestor.keys()).map(IATA =>
            context.dataLoader.location.loadById(IATA),
          ),
        );

        return locations.map(location => {
          const ancestorValue = ancestor.get(location.locationId) || {
            pickup: new Date(),
            dropoff: new Date(),
          };

          return {
            location,
            pickup: ancestorValue.pickup,
            dropoff: ancestorValue.dropoff,
          };
        });
      },
    },
  },
});
