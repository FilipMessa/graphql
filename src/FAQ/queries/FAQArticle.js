// @flow

import { GraphQLID, GraphQLNonNull } from 'graphql';
import { fromGlobalId } from 'graphql-relay';

import FAQArticle from '../types/outputs/FAQArticle';
import LanguageInput from '../../common/types/inputs/LanguageInput';
import type { GraphqlContextType } from '../../common/services/GraphqlContext';

export default {
  type: FAQArticle,
  description: 'Retrieve single FAQ article.',
  args: {
    id: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'ID of FAQ article to retrieve.',
    },
    language: {
      type: LanguageInput,
      description: 'Language in which the text of the article is returned.',
    },
  },
  resolve: async (
    ancestor: mixed,
    { language, id }: Object,
    { dataLoader }: GraphqlContextType,
  ) => {
    const { id: originalId, type } = fromGlobalId(id);

    if (type !== 'FAQArticle') {
      throw new Error(
        `FAQArticle ID mishmash. You cannot query FAQ with ID ` +
          `'${id}' because this ID is not ID of the FAQArticle. ` +
          `Please use opaque ID of the FAQArticle.`,
      );
    }

    return dataLoader.FAQArticle.load({
      originalId,
      language,
    });
  },
};
