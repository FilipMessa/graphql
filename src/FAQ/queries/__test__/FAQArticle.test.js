// @flow

import { graphql, RestApiMock } from '../../../common/services/TestingTools';
import FAQArticle from '../../datasets/FAQArticle-39.json';

describe('FAQArticle', () => {
  beforeEach(() => {
    RestApiMock.onGet(
      'https://api.skypicker.com/knowledgebase/api/v1/articles/39',
    ).replyWithData(FAQArticle);
  });

  it('should return article detail', async () => {
    const id = 'RkFRQXJ0aWNsZTozOQ==';
    const query = `
    query FAQArticle($id: ID!) {
      FAQArticle(id: $id, language: en) {
        id
        title,
        perex
        content
      }
    }
    `;
    expect(await graphql(query, { id })).toMatchSnapshot();
  });
  it('should return wrong response', async () => {
    const id = '43453';
    const query = `
        query FAQArticle($id: ID!) {
          FAQArticle(id: $id, language: en) {
            id
            title,
            perex
            content
          }
        }
        `;
    expect(await graphql(query, { id })).toMatchSnapshot();
  });
});
