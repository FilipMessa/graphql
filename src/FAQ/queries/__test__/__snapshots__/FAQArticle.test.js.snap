// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FAQArticle should return article detail 1`] = `
Object {
  "data": Object {
    "FAQArticle": Object {
      "content": "<div><strong>You may also search for your flight number with:</strong></div><ul><li><a href=\\"http://www.flightradar24.com\\">Flightradar24.com</a></li><li>Google</li></ul><div><em>Your flight number may be found on your e-ticket.</em></div>",
      "id": "RkFRQXJ0aWNsZTozOQ==",
      "originalId": 39,
      "perex": "Your departure and arrival terminals should be stated on your boarding pass.",
      "title": "How do I know which terminals I will be using?",
    },
  },
}
`;

exports[`FAQArticle should return wrong response 1`] = `
Object {
  "data": Object {
    "FAQArticle": null,
  },
  "errors": Array [
    [GraphQLError: FAQArticle ID mishmash. You cannot query FAQ with ID '43453' because this ID is not ID of the FAQArticle. Please use opaque ID of the FAQArticle.],
  ],
}
`;
